frame/staking/src/migration.rs

```diff
@@ -111,6 +114,92 @@ pub mod inner {
 		}
 	}
 
+	// migrate storage from v2 to v3:
+	//
+	// In version 2 the staking module has several issue about handling session delay.
+	// In V2 the current era is considered the active one.
+	//
+	// After the migration the current era will still be considered the active one. And the delay
+	// issue will be fixed when planning the next era.
+	//
+	// * create:
+	//   * ActiveEraStart
+	//   * ErasRewardPoints
+	//   * ActiveEra
+	//   * ErasStakers
+	//   * ErasValidatorPrefs
+	//   * ErasTotalStake
+	//   * ErasStartSessionIndex
+	// * translate StakingLedger
+	// * removal of:
+	//   * Stakers
+	//   * SlotStake
+	//   * CurrentElected
+	//   * CurrentEraStart
+	//   * CurrentEraStartSessionIndex
+	//   * CurrentEraPointsEarned
+	pub fn to_v3<T: Trait>(version: &mut VersionNumber) {
```

Comment:
is this code tested anywhere?

Reply:
yes here but I agree it is not sure this take into account all state kusama can be in. https://github.com/paritytech/substrate/pull/4474/files#diff-6031f6ad0d6b579272b1748548a364ddR2863

Reply:
there are two other tests:
- use a test environment in `master` in order to create a toy state, then dump the keys in externalities; then make a test in this branch whose environment contains only those keys and check to make sure the upgrade completes correctly.
- dumping the state of kusama (there's an rpc `state_getPairs(prefix)` to make this easy), then place it in a dev chain's genesis and test to make sure it completes successfully.